Makefile

```diff
@@ -0,0 +1,160 @@
+        CARGO_CMD_OPTS += --verbose
+    endif
+endif
+
+ifeq ($(NOVENDOR),)
+    # This MUST be kept in sync with API_HEADERS under lib.rs in libbpf-sys
+    LIBBPF_API_HEADERS := bpf.h \
+                          libbpf.h \
+                          btf.h \
+                          bpf_helpers.h \
+                          bpf_helper_defs.h \
+                          bpf_tracing.h \
+                          bpf_endian.h \
+                          bpf_core_read.h \
+                          libbpf_common.h \
+                          usdt.bpf.h
+
+    LIBBPF_SYS_LIBBPF_BASE_PATH := $(dir $(shell cargo metadata --format-version=1 | jq -r '.packages | .[] | select(.name == "libbpf-sys") | .manifest_path'))
+    LIBBPF_SYS_LIBBPF_INCLUDES :=  $(wildcard $(addprefix $(LIBBPF_SYS_LIBBPF_BASE_PATH)/libbpf/src/, $(LIBBPF_API_HEADERS)))
+    LIBBPF_INCLUDES := $(ROOT_DIR)/src/.out
+endif
+
+FILTER_INCLUDES := src/core/filters/packets/bpf/include \
+                   src/core/filters/meta/bpf/include
+INCLUDES_ALL := $(abspath $(wildcard src/core/probe/bpf/include \
+                                     src/core/probe/kernel/bpf/include \
+                                     src/core/probe/user/bpf/include \
+                                     src/core/events/bpf/include \
+                                     src/core/tracking/bpf/include \
+                                     /usr/include/x86_64-linux-gnu \
+                                     $(FILTER_INCLUDES)))
+INCLUDES_ALL += $(LIBBPF_INCLUDES)
+
+OVS_INCLUDES := $(abspath src/module/ovs/bpf/include)
+INCLUDES := $(addprefix -I, $(INCLUDES_ALL))
+
+EBPF_PROBES := $(abspath src/core/probe/kernel/bpf \
+                        src/core/probe/user/bpf)
+
+GENERIC_HOOKS := $(abspath src/module/skb/bpf \
+                           src/module/skb_drop/bpf \
+                           src/module/skb_tracking/bpf \
+                           src/module/nft/bpf \
+                           src/module/ct/bpf)
+
+OVS_HOOKS := $(abspath src/module/ovs/bpf)
+OUT_NAME := HOOK
+
+JOBS := $(patsubst -j%,%,$(filter -j%,$(MAKEFLAGS)))
+
+ifneq ($(JOBS),)
+    CARGO_JOBS := $(JOBS)
+else
+    CARGO_JOBS := 1
+endif
+
+all: build
+
+install: release
+	$(CARGO) $(CARGO_OPTS) install $(CARGO_CMD_OPTS) --path=.
+
+release: CARGO_CMD_OPTS += --release
+release: build
+
+build: ebpf
+	$(call out_console,CARGO,building retis ($(if $(findstring --release,$(CARGO_CMD_OPTS)),release,debug))...)
+	$(Q)CARGO_BUILD_JOBS=$(CARGO_JOBS) \
+	RETIS_RELEASE_VERSION=$(RELEASE_VERSION) \
+	RETIS_RELEASE_NAME=$(RELEASE_NAME) \
+	$(CARGO) $(CARGO_OPTS) build $(CARGO_CMD_OPTS)
+
+test: ebpf
+	$(call out_console,CARGO,running tests ...)
+	$(Q)CARGO_BUILD_JOBS=$(CARGO_JOBS) \
+	$(CARGO) $(CARGO_OPTS) test $(CARGO_CMD_OPTS)
+
+bench: ebpf
+	$(call out_console,CARGO,building benchmarks ...)
+	$(Q)CARGO_BUILD_JOBS=$(CARGO_JOBS) \
+	$(CARGO) $(CARGO_OPTS) build -F benchmark --release $(CARGO_CMD_OPTS)
```

Comment:
Suggestion: I think the following would remove duplication and make sure we have all set the right way for every variant; eg. release name/version or cargo jobs # would always be set — future options too.

```make
all: debug

install: release
	$(CARGO) $(CARGO_OPTS) install $(CARGO_CMD_OPTS) --path=.

define build
	$(MAKE) ebpf
	$(call out_console,CARGO,$(2)...)
	$(Q)CARGO_BUILD_JOBS=$(CARGO_JOBS) \
	RETIS_RELEASE_VERSION=$(RELEASE_VERSION) \
	RETIS_RELEASE_NAME=$(RELEASE_NAME) \
	$(CARGO) $(CARGO_OPTS) $(1) $(CARGO_CMD_OPTS)
endef

debug:
	$(call build, build, building retis (debug))

release:
	$(call build, build --release, building retis (release))

test:
	$(call build, test, building and running tests)

bench:
	$(call build, build -F benchmark --release, building benchmarks)
```

Reply:
I'm ok with factoring out some things, but I prefer to keep the prereq visible, hiding it in a macro call doesn't really make it as obvious as it should be

Reply:
Works for me, that's the part I wasn't sure about :)